# Author: Ali Pirani
configfile: "config/config.yaml"

import pandas as pd
import os

samples_df = pd.read_csv(config["samples"])
BARCODE = list(samples_df['barcode_id'])
SAMPLE = list(samples_df['sample_id'])
PREFIX = config["prefix"]
SHORTREADS = list(samples_df['sample_id'])

samples_df['combination'] = samples_df[['barcode_id', 'sample_id']].agg('/'.join, axis=1)
COMBINATION = list(samples_df['combination'])

if not os.path.exists("results/" + PREFIX):
    os.system("mkdir %s" % "results/" + PREFIX)

rule all:
    input:
        pycoqc_report = expand("results/{prefix}/pycoqc/{prefix}.html", prefix=PREFIX),
        trim_r1 =  expand("results/{prefix}/trimmomatic/{barcode}/{sample}/{sample}_R1_paired.fastq.gz", zip, barcode=BARCODE, sample=SAMPLE, prefix=PREFIX),
        # trimmed = expand("results/{prefix}/porechop/{barcode}/{sample}/{sample}.trimmed.fastq", barcode=BARCODE, sample=SAMPLE, prefix=PREFIX),
        # flye_assembly = expand("results/{prefix}/flye/{barcode}/{sample}/{sample}_flye.fasta", zip, barcode=BARCODE, sample=SAMPLE, prefix=PREFIX),
        # medaka_out = expand("results/{prefix}/medaka/{barcode}/{sample}/{sample}_medaka.fasta", zip, barcode=BARCODE, sample=SAMPLE, prefix=PREFIX),
        polypolish = expand("results/{prefix}/polypolish/{barcode}/{sample}_flye_medaka_polypolish.fasta", zip, barcode=BARCODE, sample=SAMPLE, prefix=PREFIX),
        polypolish_unicycler = expand("results/{prefix}/polypolish_unicycler/{barcode}/{sample}/{sample}_unicycler_polypolish.fasta", zip, barcode=BARCODE, sample=SAMPLE, prefix=PREFIX),
        prokka = expand("results/{prefix}/prokka/{barcode}/{sample}/", zip, barcode=BARCODE, sample=SAMPLE, prefix=PREFIX),
        quast_out = expand("results/{prefix}/quast/{barcode}/{sample}/{sample}", zip, barcode=BARCODE, sample=SAMPLE, prefix=PREFIX),
        busco_out = expand("results/{prefix}/busco/{barcode}/{sample}/{sample}_busco", zip, barcode=BARCODE, sample=SAMPLE, prefix=PREFIX),
        multiqc_out = "results/%s/multiqc" % config["prefix"]

rule pycoqc:
    input:
        seq_summary = config["long_reads"] + "/sequencing_summary.txt",
    output:
       "results/{prefix}/pycoqc/{prefix}.html",
    log:
        "logs/{prefix}/pycoqc/{prefix}.log"        
    conda:
        "envs/pycoqc.yaml"
    log:
        "logs/{prefix}/pycoqc/{barcode}/{sample}/{sample}.log"  
    shell:
       'pycoQC -f {input.seq_summary} -o {output} &>{log}'

rule trim_nano_adaptors:
    input:
        longreads = config["long_reads"] + "/{barcode}/",
        #samplename = lambda wildcards: expand(str(config["short_reads"] + "/" + f"{wildcards.sample}_L001_R1_001.fastq.gz")),
    output:
        trimmed = "results/{prefix}/porechop/{barcode}/{sample}/{sample}.trimmed.fastq"
    log:
        "logs/{prefix}/porechop/{barcode}/{sample}/{sample}.log"        
    conda:
        "envs/porechop.yaml"
    log:
        "logs/{prefix}/porechop/{barcode}/{sample}/{sample}.log"
    shell:
        'porechop -i {input.longreads} -o {output.trimmed} -t 8 --discard_middle &>{log}'

rule flye:
    input:
        trimmed = "results/{prefix}/porechop/{barcode}/{sample}/{sample}.trimmed.fastq",
    output:
        assembly = f"results/{{prefix}}/flye/{{barcode}}/{{sample}}/{{sample}}_flye.fasta",
    params:
        assembly = directory("results/{prefix}/flye/{barcode}/{sample}"),
        size = config["genome_size"],
        threads = config["threads"],
        flye_options = config["flye_options"],
        prefix = f"{{sample}}",
    log:
        "logs/{prefix}/flye/{barcode}/{sample}/{sample}.log"  
    conda:
        "envs/flye.yaml"
    shell:
        "flye --nano-hq {input.trimmed} -g {params.size} -o {params.assembly} -t {params.threads} {params.flye_options} && cp {params.assembly}/assembly.fasta {params.assembly}/{params.prefix}_flye.fasta &>{log}"


rule medaka:
    input:
        flye_assembly = "results/{prefix}/flye/{barcode}/{sample}/{sample}_flye.fasta",
        trimmed = "results/{prefix}/porechop/{barcode}/{sample}/{sample}.trimmed.fastq",
    output:
        medaka_out = f"results/{{prefix}}/medaka/{{barcode}}/{{sample}}/{{sample}}_medaka.fasta"
    params:
        medaka_out_dir = directory("results/{prefix}/medaka/{barcode}/{sample}"),
        threads = config["threads"],
        prefix = f"{{sample}}",
        model = config["medaka_model"],
    conda:
        "envs/medaka.yaml"
    log:
        "logs/{prefix}/medaka/{barcode}/{sample}/{sample}.log"
    shell:
        "medaka_consensus -i {input.trimmed} -d {input.flye_assembly} -o {params.medaka_out_dir} -t {params.threads} -m {params.model} && bwa index {params.medaka_out_dir}/consensus.fasta && cp {params.medaka_out_dir}/consensus.fasta {params.medaka_out_dir}/{params.prefix}_medaka.fasta &>{log}"


rule trimmomatic_pe:
    input:
        r1 = lambda wildcards: expand(str(config["short_reads"] + "/" + f"{wildcards.sample}_L001_R1_001.fastq.gz")),
        r2 = lambda wildcards: expand(str(config["short_reads"] + "/" + f"{wildcards.sample}_L001_R2_001.fastq.gz")),
    output:
        r1 = f"results/{{prefix}}/trimmomatic/{{barcode}}/{{sample}}/{{sample}}_R1_paired.fastq.gz",
        r2 = f"results/{{prefix}}/trimmomatic/{{barcode}}/{{sample}}/{{sample}}_R2_paired.fastq.gz", 
        # reads where trimming entirely removed the mate
        r1_unpaired = f"results/{{prefix}}/trimmomatic/{{barcode}}/{{sample}}/{{sample}}_R1_unpaired.fastq.gz",
        r2_unpaired = f"results/{{prefix}}/trimmomatic/{{barcode}}/{{sample}}/{{sample}}_R2_unpaired.fastq.gz",
    params:
        adapter_filepath=config["adapter_file"],
        seed=config["seed_mismatches"],
        palindrome_clip=config["palindrome_clipthreshold"],
        simple_clip=config["simple_clipthreshold"],
        minadapterlength=config["minadapterlength"],
        keep_both_reads=config["keep_both_reads"],
        window_size=config["window_size"],
        window_size_quality=config["window_size_quality"],
        minlength=config["minlength"],
        headcrop_length=config["headcrop_length"],
        threads = config["ncores"],
    log:
        "logs/{prefix}/trimmomatic/{barcode}/{sample}/{sample}.log"
    conda:
        "envs/trimmomatic.yaml"
    shell:
        "trimmomatic PE {input.r1} {input.r2} {output.r1} {output.r1_unpaired} {output.r2} {output.r2_unpaired} -threads {params.threads} ILLUMINACLIP:{params.adapter_filepath}:{params.seed}:{params.palindrome_clip}:{params.simple_clip}:{params.minadapterlength}:{params.keep_both_reads} SLIDINGWINDOW:{params.window_size}:{params.window_size_quality} MINLEN:{params.minlength} HEADCROP:{params.headcrop_length} &>{log}"

rule bwaalign:
    input:
        r1 = lambda wildcards: expand(f"results/{wildcards.prefix}/trimmomatic/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_R1_paired.fastq.gz"),
        r2 = lambda wildcards: expand(f"results/{wildcards.prefix}/trimmomatic/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_R2_paired.fastq.gz"),
        medaka_assembly = f"results/{{prefix}}/medaka/{{barcode}}/{{sample}}/{{sample}}_medaka.fasta"
    output:
        samout_1 = f"results/{{prefix}}/polypolish/{{barcode}}/{{sample}}/{{sample}}_1.sam",
        samout_2 = f"results/{{prefix}}/polypolish/{{barcode}}/{{sample}}/{{sample}}_2.sam",
    params:
        threads = config["ncores"],
    conda:
        "envs/bwa.yaml"
    shell:
        "bwa mem -t12 -a {input.medaka_assembly} {input.r1} > {output.samout_1} && bwa mem -t12 -a {input.medaka_assembly} {input.r2} > {output.samout_2}"

rule bwaalign_unicycler:
    input:
        r1 = lambda wildcards: expand(f"results/{wildcards.prefix}/trimmomatic/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_R1_paired.fastq.gz"),
        r2 = lambda wildcards: expand(f"results/{wildcards.prefix}/trimmomatic/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_R2_paired.fastq.gz"),
        unicycler_assembly = f"results/{{prefix}}/unicycler/{{barcode}}/{{sample}}/{{sample}}_unicycler.fasta"
    output:
        samout_1 = f"results/{{prefix}}/polypolish_unicycler/{{barcode}}/{{sample}}/{{sample}}_1.sam",
        samout_2 = f"results/{{prefix}}/polypolish_unicycler/{{barcode}}/{{sample}}/{{sample}}_2.sam",
    params:
        threads = config["ncores"],
    conda:
        "envs/bwa.yaml"
    shell:
        "bwa index {input.unicycler_assembly} && bwa mem -t12 -a {input.unicycler_assembly} {input.r1} > {output.samout_1} && bwa mem -t12 -a {input.unicycler_assembly} {input.r2} > {output.samout_2}"

rule polypolish:
    input:
        r1 = lambda wildcards: expand(f"results/{wildcards.prefix}/trimmomatic/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_R1_paired.fastq.gz"),
        r2 = lambda wildcards: expand(f"results/{wildcards.prefix}/trimmomatic/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_R2_paired.fastq.gz"),
        # medaka_assembly = f"results/medaka/{{barcode}}/consensus.fasta",
        medaka_assembly = lambda wildcards: expand(f"results/{wildcards.prefix}/medaka/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}" + "_medaka.fasta"),
        samout_1 = lambda wildcards: expand(f"results/{wildcards.prefix}/polypolish/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_1.sam"),
        samout_2 = lambda wildcards: expand(f"results/{wildcards.prefix}/polypolish/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_2.sam"),
    output:
        filtersam1 = f"results/{{prefix}}/polypolish/{{barcode}}/{{sample}}_filtered_1.sam",
        filtersam2 = f"results/{{prefix}}/polypolish/{{barcode}}/{{sample}}_filtered_2.sam",
        flye_medaka_polypolish = f"results/{{prefix}}/polypolish/{{barcode}}/{{sample}}_flye_medaka_polypolish.fasta",
    params:
        threads = config["ncores"],
    conda:
        "envs/polypolish.yaml"
    shell:
        "polypolish_insert_filter.py --in1 {input.samout_1} --in2 {input.samout_2} --out1 {output.filtersam1} --out2 {output.filtersam2}  && polypolish {input.medaka_assembly} {output.filtersam1} {output.filtersam2} > {output.flye_medaka_polypolish}"

rule unicycler:
    input:
        r1 = lambda wildcards: expand(f"results/{wildcards.prefix}/trimmomatic/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_R1_paired.fastq.gz"),
        r2 = lambda wildcards: expand(f"results/{wildcards.prefix}/trimmomatic/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_R2_paired.fastq.gz"),
        trimmed_long = "results/{prefix}/porechop/{barcode}/{sample}/{sample}.trimmed.fastq",
    output:
        unicycler_assembly = f"results/{{prefix}}/unicycler/{{barcode}}/{{sample}}/{{sample}}_unicycler.fasta",
    params:
        unicycler_assembly_out = directory("results/{{prefix}}/unicycler/{{barcode}}/{{sample}}"),
        threads = config["ncores"],
        prefix = f"{{sample}}",
    conda:
        "envs/unicycler.yaml"
    shell:
        "unicycler -1 {input.r1} -2 {input.r2} -l {input.trimmed_long} -o {params.unicycler_assembly_out} && cp {params.unicycler_assembly_out}/assembly.fasta {params.unicycler_assembly_out}/{params.prefix}_unicycler.fasta"

rule polypolish_unicycler:
    input:
        r1 = lambda wildcards: expand(f"results/{wildcards.prefix}/trimmomatic/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_R1_paired.fastq.gz"),
        r2 = lambda wildcards: expand(f"results/{wildcards.prefix}/trimmomatic/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_R2_paired.fastq.gz"),
        unicycler_assembly = f"results/{{prefix}}/unicycler/{{barcode}}/{{sample}}/{{sample}}_unicycler.fasta",
        samout_1 = lambda wildcards: expand(f"results/{wildcards.prefix}/polypolish_unicycler/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_1.sam"),
        samout_2 = lambda wildcards: expand(f"results/{wildcards.prefix}/polypolish_unicycler/" + f"{wildcards.barcode}/{wildcards.sample}/{wildcards.sample}_2.sam"),
    output:
        filtersam1 = f"results/{{prefix}}/polypolish_unicycler/{{barcode}}/{{sample}}/{{sample}}_filtered_1.sam",
        filtersam2 = f"results/{{prefix}}/polypolish_unicycler/{{barcode}}/{{sample}}/{{sample}}_filtered_2.sam",
        unicycler_polypolish = f"results/{{prefix}}/polypolish_unicycler/{{barcode}}/{{sample}}/{{sample}}_unicycler_polypolish.fasta",
    params:
        threads = config["ncores"],
    conda:
        "envs/polypolish.yaml"
    shell:
        "polypolish_insert_filter.py --in1 {input.samout_1} --in2 {input.samout_2} --out1 {output.filtersam1} --out2 {output.filtersam2}  && polypolish {input.unicycler_assembly} {output.filtersam1} {output.filtersam2} > {output.unicycler_polypolish}"

rule prokka:
    input:
        unicycler_polypolish = f"results/{{prefix}}/polypolish_unicycler/{{barcode}}/{{sample}}/{{sample}}_unicycler_polypolish.fasta",
        unicycler_assembly = f"results/{{prefix}}/unicycler/{{barcode}}/{{sample}}/{{sample}}_unicycler.fasta",
        flye_medaka_polypolish = f"results/{{prefix}}/polypolish/{{barcode}}/{{sample}}_flye_medaka_polypolish.fasta",
        medaka_assembly = f"results/{{prefix}}/medaka/{{barcode}}/{{sample}}/{{sample}}_medaka.fasta",
        flye_assembly = f"results/{{prefix}}/flye/{{barcode}}/{{sample}}/{{sample}}_flye.fasta",
    output:
        #unicycler_annotation = f"results/{{prefix}}/prokka/{{barcode}}/{{sample}}/{{sample}}_unicycler/{{sample}}_unicycler.gff",
        prokka_dir_out = directory("results/{prefix}/prokka/{barcode}/{sample}/"),
    params:
        threads = config["ncores"],
        prefix = f"{{sample}}",
        options = config["prokka_options"],
        prokka_dir = directory("results/{prefix}/prokka/{barcode}/{sample}/{sample}"),
    conda:
        "envs/prokka.yaml"
    shell:
        "prokka {params.options} --strain {params.prefix} -outdir {params.prokka_dir}_unicycler -prefix {params.prefix}_unicycler {input.unicycler_assembly} && prokka {params.options} --strain {params.prefix} -outdir {params.prokka_dir}_flye -prefix {params.prefix}_flye {input.flye_assembly} && prokka {params.options} --strain {params.prefix} -outdir {params.prokka_dir}_medaka -prefix {params.prefix}_medaka {input.medaka_assembly} && prokka {params.options} --strain {params.prefix} -outdir {params.prokka_dir}_flye_medaka_polypolish -prefix {params.prefix}_flye_medaka_polypolish {input.flye_medaka_polypolish} && prokka {params.options} --strain {params.prefix} -outdir {params.prokka_dir}_unicycler_polypolish -prefix {params.prefix}_unicycler_polypolish {input.unicycler_polypolish}"


rule quast:
    input:
        unicycler_polypolish = f"results/{{prefix}}/polypolish_unicycler/{{barcode}}/{{sample}}/{{sample}}_unicycler_polypolish.fasta",
        unicycler_assembly = f"results/{{prefix}}/unicycler/{{barcode}}/{{sample}}/{{sample}}_unicycler.fasta",
        flye_medaka_polypolish = f"results/{{prefix}}/polypolish/{{barcode}}/{{sample}}_flye_medaka_polypolish.fasta",
        medaka_assembly = f"results/{{prefix}}/medaka/{{barcode}}/{{sample}}/{{sample}}_medaka.fasta",
        flye_assembly = f"results/{{prefix}}/flye/{{barcode}}/{{sample}}/{{sample}}_flye.fasta",
    output:
        quast_out = f"results/{{prefix}}/quast/{{barcode}}/{{sample}}/{{sample}}",
    params:
        threads = config["ncores"],
        quast_dir = directory("results/{{prefix}}/quast/{barcode}/{sample}"),
    conda:
        "envs/quast.yaml"
    shell:
        "quast {input.unicycler_assembly} -o {params.quast_dir}_unicycler && quast {input.medaka_assembly} -o {params.quast_dir}_medaka && quast {input.unicycler_polypolish} -o {params.quast_dir}_unicycler_polypolish && quast {input.flye_medaka_polypolish} -o {params.quast_dir}_flye_medaka_polypolish && quast {input.flye_assembly} -o {params.quast_dir}_flye"


rule busco:
    input:
        unicycler_polypolish = f"results/{{prefix}}/polypolish_unicycler/{{barcode}}/{{sample}}/{{sample}}_unicycler_polypolish.fasta",
        unicycler_assembly = f"results/{{prefix}}/unicycler/{{barcode}}/{{sample}}/{{sample}}_unicycler.fasta",
        flye_medaka_polypolish = f"results/{{prefix}}/polypolish/{{barcode}}/{{sample}}_flye_medaka_polypolish.fasta",
        medaka_assembly = f"results/{{prefix}}/medaka/{{barcode}}/{{sample}}/{{sample}}_medaka.fasta",
        flye_assembly = f"results/{{prefix}}/flye/{{barcode}}/{{sample}}/{{sample}}_flye.fasta",
    output:
        busco_out = directory("results/{prefix}/busco/{barcode}/{sample}/{sample}_busco/"),
    params:
        threads = config["ncores"],
    conda:
        "envs/busco.yaml"
    shell:
        "busco -i {input.unicycler_assembly} -m genome -l bacteria_odb10 -o {output.busco_out}"

rule multiqc:
    input:
        prokka_dir_out = f"results/{{prefix}}/prokka/%s/%s" % (BARCODE[0], SAMPLE[0]),
        busco_dir_out = f"results/{{prefix}}/busco/%s/%s/%s_busco" % (BARCODE[0], SAMPLE[0], SAMPLE[0]),
        quast_dir_out = f"results/{{prefix}}/quast/%s/%s/%s" % (BARCODE[0], SAMPLE[0], SAMPLE[0]),
        pycoqc_dir_out = f"results/{{prefix}}/pycoqc/{{prefix}}.html",

    output:
        multiqc_out = directory("results/{prefix}/multiqc"),
    params:
        multiqc_out_dir = directory("results/{prefix}/multiqc/"),
        threads = config["ncores"],
    priority: 100000    
    conda:
        "envs/multiqc.yaml"
    shell:
        "multiqc -o {params.multiqc_out_dir} -f {input.prokka_dir_out} {input.busco_dir_out} {input.quast_dir_out} {input.pycoqc_dir_out}"